src/analysis.rs

```diff
@@ -0,0 +1,547 @@
+        let e3 = Parser::new().parse("(?:m)").unwrap();
+
+        assert!(is_intersecting_fset(&e1, &e2));
+        assert!(!is_intersecting_fset(&e1, &e3));
+    }
+
+    #[test]
+    fn fset_concat() {
+        let e1 = Parser::new().parse("aa(?:nb)").unwrap();
+        let e2 = Parser::new().parse("aa(?:rq)").unwrap();
+        let e3 = Parser::new().parse("bb(?:m)").unwrap();
+
+        assert!(is_intersecting_fset(&e1, &e2));
+        assert!(!is_intersecting_fset(&e1, &e3));
+    }
+
+    #[test]
+    fn fset_word_boundary_dropped() {
+        let e1 = Parser::new().parse(r"aa").unwrap();
+        let e2 = Parser::new().parse(r"\baa").unwrap();
+        let e3 = Parser::new().parse(r"\bbb").unwrap();
+
+        assert!(is_intersecting_fset(&e1, &e2));
+        assert!(!is_intersecting_fset(&e1, &e3));
+    }
+
+    #[test]
+    fn fset_word_boundary_all() {
+        let e1 = Parser::new().parse(r"aa").unwrap();
+        let e2 = Parser::new().parse(r"\b").unwrap();
+
+        assert!(is_intersecting_fset(&e1, &e2));
+    }
+
+    #[test]
+    fn fset_not_word_boundary_dropped() {
+        let e1 = Parser::new().parse(r"aa").unwrap();
+        let e2 = Parser::new().parse(r"\Baa").unwrap();
+        let e3 = Parser::new().parse(r"\Bbb").unwrap();
+
+        assert!(is_intersecting_fset(&e1, &e2));
+        assert!(!is_intersecting_fset(&e1, &e3));
+    }
+
+    #[test]
+    fn fset_not_word_boundary_all() {
+        let e1 = Parser::new().parse(r"aa").unwrap();
+        let e2 = Parser::new().parse(r"\B").unwrap();
+
+        assert!(is_intersecting_fset(&e1, &e2));
+    }
+
+    #[test]
+    fn fset_start_anchor_dropped() {
+        let e1 = Parser::new().parse(r"aa").unwrap();
+        let e2 = Parser::new().parse(r"^aa").unwrap();
+        let e3 = Parser::new().parse(r"^bb").unwrap();
+
+        assert!(is_intersecting_fset(&e1, &e2));
+        assert!(!is_intersecting_fset(&e1, &e3));
+    }
+
+    #[test]
+    fn fset_terminal_emptylook_all() {
+        let e = Parser::new().parse(r"a*\b").unwrap();
+
+        let mut total_accept = FirstSet::anychar();
+        total_accept.accepts_empty = true;
+
+        assert_eq!(total_accept, fset_of(&e));
+    }
+
+    #[test]
+    fn fset_empty_alt() {
+        let e1 = Parser::new().parse(r"(?:a|())b").unwrap();
+        let e2 = Parser::new().parse(r"b").unwrap();
+
+        assert!(is_intersecting_fset(&e1, &e2));
+    }
+
+    //
+    // onepass smoke tests
+    //
+
+    // This test is pulled right from some of Russ Cox's
+    // comments on onepass regex.
+    //
+    // Note that Russ Cox's other example of a onepass regex
+    // (r"(\d+)-(\d+)") is actually not onepass for us because
+    // there is byte-level nondeterminism in the \d character
+    // class, and we care about things in the byte space rather
+    // than the character space. If you do a onepass engine at
+    // the character level, Cox's example is indeed onepass.
```

Comment:
Could we add a test for said regex?

IIRC, RE2 uses a byte level representation as well. I wonder if the real reason here is that `\d` is Unicode aware by default here, but is only ASCII aware in RE2. That is, would `(?-u)(\d+)-(\d+)` be one pass?

I think overall I would expect to see quite a bit more tests that check the top level `is_onepass` routine here. :-)

Reply:
Yeah, the RE2 onepass matcher is definitely byte oriented because RE2 doesn't have any character level byte codes. Even in this crate, the pike VM can match Unicode character classes without character level byte codes. For RE2, see: https://github.com/google/re2/blob/e6acac839be126c8a20a20183f20a23fa9f4895e/re2/onepass.cc#L459-L524

The Go onepass matcher---along with all of the matchers in Go---are however character oriented. There is no byte oriented matching in Go's regexp engine IIRC.